From 2567c94e0a183a65a7f3a8680f4e37b2d72406a5 Mon Sep 17 00:00:00 2001
From: CL Wang <cl634@andestech.com>
Date: Wed, 5 Mar 2025 11:42:25 +0800
Subject: [PATCH 165/168] The FTSSDP010 DMA driver init is adjusted to occur
 after ATCDMAC300(#261)

Due to the change in the ATCDMAC300 DMA engine driver's initialization
order 'commit 4267e264d5 ("[backport] Support AIA extension on Andes
ae350 platform (#240)")', the FTSSP010 DMA driver must be initialized
after ATCDMAC300 to ensure proper DMA functionality.

Reviewed-on: https://gitea.andestech.com/RD-SW/linux/pulls/261
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Charles Ci-Jyun Wu <dminus@andestech.com>
Reviewed-by: randolph <randolph@andestech.com>
Co-authored-by: CL Wang <cl634@andestech.com>
Co-committed-by: CL Wang <cl634@andestech.com>
---
 sound/v5/ftssp010_alsa.c |  5 +++
 sound/v5/ftssp010_dma.c  |  9 ------
 sound/v5/ftssp010_dma.h  | 68 +++++++++++++++++++---------------------
 3 files changed, 37 insertions(+), 45 deletions(-)

diff --git a/sound/v5/ftssp010_alsa.c b/sound/v5/ftssp010_alsa.c
index 3ad741e7bdea..21173002cfa1 100644
--- a/sound/v5/ftssp010_alsa.c
+++ b/sound/v5/ftssp010_alsa.c
@@ -1433,11 +1433,16 @@ static int atf_ac97_probe(struct platform_device *pdev)
 		return -ENXIO;
 	}
 
+	ret = dmad_init();
+	if (ret)
+		return ret;
+
 	return ftssp_alsa_init(pdev);
 }
 
 static int atf_ac97_remove(struct platform_device *pdev)
 {
+	dmad_release();
 	return ftssp_alsa_i2c_i2s_exit();
 }
 
diff --git a/sound/v5/ftssp010_dma.c b/sound/v5/ftssp010_dma.c
index 8d7b0fc258d4..2363606eabac 100644
--- a/sound/v5/ftssp010_dma.c
+++ b/sound/v5/ftssp010_dma.c
@@ -2342,12 +2342,3 @@ int dmad_release(void)
 
 	return 0;
 }
-
-subsys_initcall(dmad_init);
-
-static void __exit dmad_exit(void)
-{
-	dmad_release();
-}
-
-module_exit(dmad_exit);
diff --git a/sound/v5/ftssp010_dma.h b/sound/v5/ftssp010_dma.h
index 1261d9ed17d2..08fd546d6313 100644
--- a/sound/v5/ftssp010_dma.h
+++ b/sound/v5/ftssp010_dma.h
@@ -410,44 +410,40 @@ typedef struct dmad_drb {
 
 } dmad_drb;
 
-/******************************************************************************
- * DMAD Driver Interface
- ******************************************************************************/
-extern int dmad_channel_alloc(dmad_chreq *ch_req);
-extern int dmad_channel_free(dmad_chreq *ch_req);
-extern int dmad_channel_enable(const dmad_chreq *ch_req, u8 enable);
-extern u32 dmad_max_size_per_drb(dmad_chreq *ch_req);
-extern u32 dmad_bytes_to_cycles(dmad_chreq *ch_req, u32 byte_size);
-
-extern int dmad_kickoff_requests(dmad_chreq *ch_req);
-extern int dmad_drain_requests(dmad_chreq *ch_req, u8 shutdown);
-
-/* for performance reason, these two functions are platform-specific */
-extern int dmad_probe_irq_source_ahb(void);
-/* note: hw_ptr here is phyical address of dma source or destination */
-extern dma_addr_t dmad_probe_hw_ptr_src(dmad_chreq *ch_req);
-extern dma_addr_t dmad_probe_hw_ptr_dst(dmad_chreq *ch_req);
+/* DMAD Driver Interface */
+int dmad_init(void);
+int dmad_release(void);
+int dmad_channel_alloc(dmad_chreq *ch_req);
+int dmad_channel_free(dmad_chreq *ch_req);
+int dmad_channel_enable(const dmad_chreq *ch_req, u8 enable);
+u32 dmad_max_size_per_drb(dmad_chreq *ch_req);
+u32 dmad_bytes_to_cycles(dmad_chreq *ch_req, u32 byte_size);
+int dmad_kickoff_requests(dmad_chreq *ch_req);
+int dmad_drain_requests(dmad_chreq *ch_req, u8 shutdown);
+
+/* For performance reasons, these two functions are platform specific. */
+int dmad_probe_irq_source_ahb(void);
+
+/* Note: hw_ptr here is phyical address of dma source or destination. */
+dma_addr_t dmad_probe_hw_ptr_src(dmad_chreq *ch_req);
+dma_addr_t dmad_probe_hw_ptr_dst(dmad_chreq *ch_req);
+
+/* Routines are only valid in discrete (non-ring) mode */
+int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir, dmad_drb *drb);
+int dmad_alloc_drb(dmad_chreq *ch_req, dmad_drb **drb);
+int dmad_free_drb(dmad_chreq *ch_req, dmad_drb *drb);
+int dmad_submit_request(dmad_chreq *ch_req, dmad_drb *drb, u8 keep_fired);
+int dmad_withdraw_request(dmad_chreq *ch_req, dmad_drb *drb);
 
-/*****************************************************************************
- * routines only valid in discrete (non-ring) mode
- */
-extern int dmad_config_channel_dir(dmad_chreq *ch_req, u8 dir, dmad_drb *drb);
-extern int dmad_alloc_drb(dmad_chreq *ch_req, dmad_drb **drb);
-extern int dmad_free_drb(dmad_chreq *ch_req, dmad_drb *drb);
-extern int dmad_submit_request(dmad_chreq *ch_req,
-			       dmad_drb *drb, u8 keep_fired);
-extern int dmad_withdraw_request(dmad_chreq *ch_req, dmad_drb *drb);
-/****************************************************************************/
-
-/*****************************************************************************
- * routines only valid in ring mode
- * note: sw_ptr and hw_ptr are values offset from the ring buffer base
- *       unit of sw_ptr is data-width
- *       unit of hw_ptr returned is byte
+/*
+ * Routines are only valid in ring mode
+ * Note: sw_ptr and hw_ptr are values offset from the ring buffer base
+ *       - Unit of sw_ptr is data width
+ *       - Unit of hw_ptr returned is byte
  */
 int dmad_update_ring(dmad_chreq *ch_req);
-int dmad_update_ring_sw_ptr(dmad_chreq *ch_req,
-				   dma_addr_t sw_ptr, u8 keep_fired);
+int dmad_update_ring_sw_ptr(dmad_chreq *ch_req, dma_addr_t sw_ptr,
+			    u8 keep_fired);
 dma_addr_t dmad_probe_ring_hw_ptr(dmad_chreq *ch_req);
 
-#endif  /* __NDS_DMAD_ATF_INC__ */
+#endif
-- 
2.34.1

